### ISSUE TITLE
Improve our test coverage

### ISSUE THREAD

**Issue by tianyizheng02** [TheAlgorithms/Python#9943]
### Feature description

Many of our existing algorithm files have little to no unit testing. This is problematic because this can easily let bugs slip through. We want some assurance that the code we currently have is correct and functional. We welcome all contributors to open PRs to help us add tests to our codebase.

### How to find low-coverage files

Go to the Actions tab in this repository and find the most recent **build** workflow run. Open the logs under "Run Tests" and scroll down until you find the section on code coverage:
```
---------- coverage: platform linux, python 3.12.0-final-0 -----------
Name                                                                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------------------
quantum/q_fourier_transform.py                                                   30     30     0%   14-93
scripts/validate_solutions.py                                                    54     54     0%   2-94
strings/min_cost_string_conversion.py                                            78     75     4%   20-57, 61-75, 79-129
...
```
The "Cover" column tells you what percentage of the lines in that file are covered by tests. We want to increase this percentage for existing files. Find a file with low coverage percentage that you wish to write tests for, add doctests for each function, and open a PR with your changes. You do not need to have a perfect coverage percentage, but all functions should have doctests.

Some files will naturally be hard to write tests for. For example, the file may be poorly written because they lack any functions. Other files might be how-tos, meaning they simply demonstrate how to use an existing library's functions rather than implementing the algorithm themselves. Ignore these kinds of files, as they will need to be rewritten eventually. Furthermore, ignore files in the `web_programming` and `project_euler` directories. Web programming files are inherently hard to test and Project Euler files have their own validation workflow, so don't worry about their test coverage.

_**When you open your PR, put "Contributes to #9943" in the PR description.**_ Do not use the word "fixes", "resolves", or "closes". This issue is an ongoing one, and your PR will not single-handedly resolve this issue.

### How to add doctests

A doctest is a unit test that is contained within the documentation comment (docstring) for a function. Here is an example of what doctests look like within a docstring:
```py
def add(a: int, b: int) -> int:
    """
    Adds two non-negative numbers.
    >>> add(1, 1)
    2
    >>> add(2, 5)
    7
    >>> add(1, 0)
    1
    >>> add(-1, -1)
    Traceback (most recent last):
    ...
    ValueError: Numbers must be non-negative
    """
```
For every function in the file you choose, you should write doctests like the ones shown above in its docstring. If a function doesn't have a docstring, add one. Your doctests should be comprehensive but not excessive: you should write just enough tests to cover all basic cases as well as all edge cases (e.g., negative numbers, empty lists, etc).

Do not simply run a function on some example inputs and put its output as the expected output for a doctest. This assumes that the function is implemented correctly when it might not be. Verify independently that your doctests and their expected outputs are correct. **Your PR will not be merged if it has failing tests.** If you happen to discover a bug while writing doctests, please fix it.

_**Please read our [contributing guidelines](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) before you contribute.**_

**Comment by SaiHarshaK** [comment]
I'll work on the palindrome script

**Comment by AnshuSharma111** [comment]
Hi. I am new and would like to contribute to this issue

**Comment by SaiHarshaK** [comment]
Please check #10135  and #10156

**Comment by AasheeshLikePanner** [comment]
Can you please review this #10154

**Comment by AasheeshLikePanner** [comment]
@tianyizheng02 Can you please review this [PR](https://github.com/TheAlgorithms/Python/pull/10194)

**Comment by AasheeshLikePanner** [comment]
@tianyizheng02 Can you please Review this two PRs

> [Adding doctests in simpson_rule.py](https://github.com/TheAlgorithms/Python/pull/10269)

> [Adding doctests in longest_sub_array.py](https://github.com/TheAlgorithms/Python/pull/10194)

**Comment by harjotjuneja6** [comment]
Can we close this issue because it is already merged.

**Comment by tianyizheng02** [comment]
No, improving test coverage is an _ongoing_ issue. It can't be resolved by a single PR because there are many, many files that need tests. Right now, it definitely hasn't been resolved.

**Comment by AasheeshLikePanner** [comment]
@tianyizheng02 Can you please review this PR 

- #10540

**Comment by SaiHarshaK** [comment]
Can we ask people to use "Ref" instead "Fixes", so that we don't have keep reopening this issue?

**Comment by RaymondDashWu** [comment]
How is the percentage determined? The tests I added only bumped the coverage by 1%

**Comment by Cheerio-ban** [comment]
Added doctest for swap_nodes .py under linkedlist data structure
https://github.com/TheAlgorithms/Python/pull/10673/commits

**Comment by tianyizheng02** [comment]
> How is the percentage determined? The tests I added only bumped the coverage by 1%

@RaymondDashWu The coverage percentage is calculated from the total number of lines covered by tests across _all_ files in the repo. As of writing, the latest builds show the following:
```
-----------------------------------------------------------------------------------------------------------
TOTAL                                                                         31846   9986    69%
```
There are 9986 lines of code in total _not_ covered by tests compared to 31846 lines of code in total, which is how the 69% comes about. The tests that you added only improved the coverage for _one_ file out of the hundreds of files that we have, and that's why it barely changed the percentage at all.

This is why I said in my previous comment that this is an _ongoing_ issue. No single PR will improve our coverage enough to "fix" this issue.

**Comment by RaymondDashWu** [comment]
> d by tests compared to 31846 lines of code in total, which is how the 69% comes about. The tests that you added only improved the coverage for _one_ file out of the hundreds of files that we have, and that's why it barely changed the percentage at all.

Thanks for this info @tianyizheng02! I figured out that I've actually been looking at the wrong percentages for test coverage. I was looking at the loading of the test session percentages thinking that's what you pointed out in the first screenshot

<img width="591" alt="image" src="https://github.com/TheAlgorithms/Python/assets/33266041/2ceced9c-4a10-4cf0-ab3c-5e6e3f4ccd46">

**Comment by nkstonks** [comment]
Hi, I was looking at the Fibonacci function, and I found an untested function "fib_recursive_term" that was nested inside of the fib_recursive function. I added tests for it, so in terms of this issue it's all good now.

But I was just wondering whether or not I should define that function globally because I don't see the point of having it nested in the fib_recursive function. 

So I want to propose to move the fib_recursive_term out from the fib_recursive function and have it as a separate function above it, would there be any issue if that happens?

**Comment by SaiHarshaK** [comment]
Maybe you can create a new function for it since the scope of this issue
seems to be only addition of do tests?

**Comment by PreciousJac0b** [comment]
@tianyizheng02 Please help review my pull request so I can write more tests for this repo.
https://github.com/TheAlgorithms/Python/pull/10751

**Comment by nkstonks** [comment]
@SaiHarshaK alright, ill make a new issue for the nested functions thing, meanwhile I'll make a pull request where I've implemented doctests to several functions once I double check everything

**Comment by imSanko** [comment]
@cclauss Now we need to work to FIX the failing TESTS for improving, while its of #10822

**Comment by vanlucard** [comment]
It's like Fourier Sistem i thought

**Comment by Saty70** [comment]
hey I want add some tests to the algorithm can guide the prerequisite that I need to know to add test cases 
will it be implemented by Pytest or something else

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/blob/1868c0b6375188a9034478a2711e40c343d00c2e/.github/workflows/build.yml#L27-L32

**Comment by HersiYussuf** [comment]
Hey! I am new to opensource I would **Love** to contribute, any tips and ideas on how can I safely and successfully contribute

**Comment by Argyrhspet1** [comment]
Hello everyone,I am getting started with open source and I would like to work on this issue .Can someone guide me on what to do?Thank you

**Comment by Amul-Sharma-404** [comment]
for the coverage report section:

---------- coverage: platform linux, python 3.12.0-final-0 -----------
Name                                                                          Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------------------
quantum/q_fourier_transform.py                                                   30     30     0%   14-93
scripts/validate_solutions.py                                                    54     54     0%   2-94
strings/min_cost_string_conversion.py                                            78     75     4%   20-57, 61-75, 79-129
...


# code for the how to add doctests
def function_name(param1: type, param2: type) -> return_type:

      # Description of what the function does.
    
      # function_name(input1, input2)
        # expected_output

    # Add more tests to cover different cases
     #   function_name(other_input1, other_input2)
    #   expected_output
    

    """

**Comment by Akash-Tiwari-890** [comment]
def add(a, b):
    """
    Adds two numbers and returns the result.

    >>> add(2, 3)
    5
    >>> add(-1, 1)
    0
    """
    return a + b
we can use this also

**Comment by cclauss** [comment]
@Akash-Tiwari-890 We do in 800+ files...
* https://github.com/search?q=repo%3ATheAlgorithms%2FPython%20%3E%3E%3E&type=code

**Comment by DhanushA1307** [comment]
Hi. I am new and would like to contribute to this issue @tianyizheng02

**Comment by HersiYussuf** [comment]
Okay have at it.

**Comment by mustalidhanerawala** [comment]
Hello, I am new over here, Is this issue still open?

**Comment by tianyizheng02** [comment]
Yes, this issue is still open.

@DhanushA1307 @mustalidhanerawala We don't assign issues in this repo. If you want to work on it, just open a PR with your changes. Our contributing guidelines already say this, so read those first if you haven't already.

**Comment by vanton** [comment]
Any new progress?

**Comment by saiteja-yadav** [comment]
Can assign me this. .
I can resolve this...

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/issues/9943#issuecomment-2389369722

**Comment by Kishore0-0-7** [comment]
> ### Feature description
> Many of our existing algorithm files have little to no unit testing. This is problematic because this can easily let bugs slip through. We want some assurance that the code we currently have is correct and functional. We welcome all contributors to open PRs to help us add tests to our codebase.
> 
> ### How to find low-coverage files
> Go to the Actions tab in this repository and find the most recent **build** workflow run. Open the logs under "Run Tests" and scroll down until you find the section on code coverage:
> 
> ```
> ---------- coverage: platform linux, python 3.12.0-final-0 -----------
> Name                                                                          Stmts   Miss  Cover   Missing
> -----------------------------------------------------------------------------------------------------------
> quantum/q_fourier_transform.py                                                   30     30     0%   14-93
> scripts/validate_solutions.py                                                    54     54     0%   2-94
> strings/min_cost_string_conversion.py                                            78     75     4%   20-57, 61-75, 79-129
> ...
> ```
> 
> The "Cover" column tells you what percentage of the lines in that file are covered by tests. We want to increase this percentage for existing files. Find a file with low coverage percentage that you wish to write tests for, add doctests for each function, and open a PR with your changes. You do not need to have a perfect coverage percentage, but all functions should have doctests.
> 
> Some files will naturally be hard to write tests for. For example, the file may be poorly written because they lack any functions. Other files might be how-tos, meaning they simply demonstrate how to use an existing library's functions rather than implementing the algorithm themselves. Ignore these kinds of files, as they will need to be rewritten eventually. Furthermore, ignore files in the `web_programming` and `project_euler` directories. Web programming files are inherently hard to test and Project Euler files have their own validation workflow, so don't worry about their test coverage.
> 
> _**When you open your PR, put "Contributes to #9943" in the PR description.**_ Do not use the word "fixes", "resolves", or "closes". This issue is an ongoing one, and your PR will not single-handedly resolve this issue.
> 
> ### How to add doctests
> A doctest is a unit test that is contained within the documentation comment (docstring) for a function. Here is an example of what doctests look like within a docstring:
> 
> ```python
> def add(a: int, b: int) -> int:
>     """
>     Adds two non-negative numbers.
>     >>> add(1, 1)
>     2
>     >>> add(2, 5)
>     7
>     >>> add(1, 0)
>     1
>     >>> add(-1, -1)
>     Traceback (most recent last):
>     ...
>     ValueError: Numbers must be non-negative
>     """
> ```
> 
> For every function in the file you choose, you should write doctests like the ones shown above in its docstring. If a function doesn't have a docstring, add one. Your doctests should be comprehensive but not excessive: you should write just enough tests to cover all basic cases as well as all edge cases (e.g., negative numbers, empty lists, etc).
> 
> Do not simply run a function on some example inputs and put its output as the expected output for a doctest. This assumes that the function is implemented correctly when it might not be. Verify independently that your doctests and their expected outputs are correct. **Your PR will not be merged if it has failing tests.** If you happen to discover a bug while writing doctests, please fix it.
> 
> _**Please read our [contributing guidelines](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) before you contribute.**_

i think i can make it :)

**Comment by Kishore0-0-7** [comment]
i think i cam make it:)

**Comment by saiteja-yadav** [comment]
> ### Feature description
> Many of our existing algorithm files have little to no unit testing. This is problematic because this can easily let bugs slip through. We want some assurance that the code we currently have is correct and functional. We welcome all contributors to open PRs to help us add tests to our codebase.
> 
> ### How to find low-coverage files
> Go to the Actions tab in this repository and find the most recent **build** workflow run. Open the logs under "Run Tests" and scroll down until you find the section on code coverage:
> 
> ```
> ---------- coverage: platform linux, python 3.12.0-final-0 -----------
> Name                                                                          Stmts   Miss  Cover   Missing
> -----------------------------------------------------------------------------------------------------------
> quantum/q_fourier_transform.py                                                   30     30     0%   14-93
> scripts/validate_solutions.py                                                    54     54     0%   2-94
> strings/min_cost_string_conversion.py                                            78     75     4%   20-57, 61-75, 79-129
> ...
> ```
> 
> The "Cover" column tells you what percentage of the lines in that file are covered by tests. We want to increase this percentage for existing files. Find a file with low coverage percentage that you wish to write tests for, add doctests for each function, and open a PR with your changes. You do not need to have a perfect coverage percentage, but all functions should have doctests.
> 
> Some files will naturally be hard to write tests for. For example, the file may be poorly written because they lack any functions. Other files might be how-tos, meaning they simply demonstrate how to use an existing library's functions rather than implementing the algorithm themselves. Ignore these kinds of files, as they will need to be rewritten eventually. Furthermore, ignore files in the `web_programming` and `project_euler` directories. Web programming files are inherently hard to test and Project Euler files have their own validation workflow, so don't worry about their test coverage.
> 
> _**When you open your PR, put "Contributes to #9943" in the PR description.**_ Do not use the word "fixes", "resolves", or "closes". This issue is an ongoing one, and your PR will not single-handedly resolve this issue.
> 
> ### How to add doctests
> A doctest is a unit test that is contained within the documentation comment (docstring) for a function. Here is an example of what doctests look like within a docstring:
> 
> ```python
> def add(a: int, b: int) -> int:
>     """
>     Adds two non-negative numbers.
>     >>> add(1, 1)
>     2
>     >>> add(2, 5)
>     7
>     >>> add(1, 0)
>     1
>     >>> add(-1, -1)
>     Traceback (most recent last):
>     ...
>     ValueError: Numbers must be non-negative
>     """
> ```
> 
> For every function in the file you choose, you should write doctests like the ones shown above in its docstring. If a function doesn't have a docstring, add one. Your doctests should be comprehensive but not excessive: you should write just enough tests to cover all basic cases as well as all edge cases (e.g., negative numbers, empty lists, etc).
> 
> Do not simply run a function on some example inputs and put its output as the expected output for a doctest. This assumes that the function is implemented correctly when it might not be. Verify independently that your doctests and their expected outputs are correct. **Your PR will not be merged if it has failing tests.** If you happen to discover a bug while writing doctests, please fix it.
> 
> _**Please read our [contributing guidelines](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) before you contribute.**_

I just wanted to say that I can resolve this...

**Comment by Cfengsu2002** [comment]
I have contributed to this issue by adding doctests for minimum_coin_change.py and improving the function's efficiency.

Added doctests to ensure proper validation of the function.
Optimized code by removing unnecessary int() conversions.
Updated function parameter type from value: str → value: int to avoid redundant type casting.
Ensured correctness by covering basic and edge cases in the doctests.
and get rid of the grammer mistakes

Please review and merge when possible. Thank you! 😊

**Comment by oyshisarker2001** [comment]
is this issue still persisting ?? . Can  I start contributing ??

**Comment by Yashcodes04** [comment]
Hello! I'm new to open source and eager to contribute. I'm interested in this issue and would like to work on it. Is it still open for contributions? Thank you!

**Comment by Adagedo** [comment]
I'm a python developer with experience in developing test cases. I have contributed to popular python frameworks like django, tensor flow and pytorch. 
I'm confidence I will develop robust, maintainable, and scalable test cases.

**Comment by Akshit2301** [comment]
Hi! I'd like to work on this issue as a first-time contributor. Please assign it to me.

**Comment by cclauss** [comment]
https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#issues

**Comment by Nitin-Prata** [comment]
Hi! I'm Nitin Pratap Singh, a Python developer exploring open source. I'd love to contribute to this issue by adding doctests to one of the low-coverage files. Could you please guide me if any file is currently unassigned? Thanks!

**Comment by cclauss** [comment]
Line 900 of step 6 starts a list of the test coverage of all algorithms.
https://github.com/TheAlgorithms/Python/actions/runs/16183259890/job/45684103076#step:6:901

**Comment by Nitin-Prata** [comment]
Thanks for the pointer,I'll go through the coverage report and pick
another file to improve test coverage.

**Comment by johnslavik** [comment]
Related: https://github.com/python/cpython/issues/44538

**Comment by Amogh2811** [comment]
Hi! I'm Amogh, a developer exploring open source. I'd love to contribute to this issue . Could you please guide me if any file is currently unassigned? Thanks!

**Comment by luca-c-coll** [comment]
@tianyizheng02  Are there additional beginner-friendly first issues, similar to this one, that I could contribute to?

**Comment by PranavTamada** [comment]
Hi maintainers, I’ve raised a PR #13036 ,related to this issue as part of Hacktoberfest 2025. If it looks good, I’d be grateful if you could review and merge it. If merging isn’t possible right now, could you please add the “hacktoberfest-accepted” label to the PR so it can count towards my Hacktoberfest contributions? Thanks a lot for your time and support!

**Comment by PranavTamada** [comment]
Hello @cclauss  @MaximSmolskiy , 

I’ve added doctests to the Dijkstra2 algorithm and raised a PR — #13036.

If everything looks good, please consider merging it or tagging it as hacktoberfest-accepted.
Thank you for your time and effort in reviewing! 🙌

**Comment by cclauss** [comment]
Will do after the 438 pull requests are reviewed.

**Comment by Akhila21-6** [comment]
Hi, I would like to add tests for the strings/min_cost_string_conversion.py file.

**Comment by frxdude** [comment]
Hi maintainers,

I would like to contribute to this issue by adding comprehensive doctests to improve the test coverage of the repository. I've reviewed the contribution guidelines and understand the requirements for adding doctests.
Could I please be assigned to work on this? I'm excited to contribute to this project and help improve the codebase quality.

Thank you!

**Comment by cclauss** [comment]
As discussed in CONTRIBUTING.md, we do not assign issues in this repository.

**Comment by vubangsi** [comment]
Thank you for reporting this issue! 

This looks like a great contribution opportunity. I noticed this repository is looking for help with $good first issue, help wanted issues.

As an automated agent, I'm here to help the open source community. This issue seems well-suited for community contributions.

If you'd like, I can help:
- Review the issue requirements
- Suggest potential solutions
- Point to relevant documentation

Would you be interested in working on this?

**Comment by vubangsi** [comment]
Thank you for reporting this issue! 

This looks like a great contribution opportunity. I noticed this repository is looking for help with $good first issue, help wanted issues.

As an automated agent, I'm here to help the open source community. This issue seems well-suited for community contributions.

If you'd like, I can help:
- Review the issue requirements
- Suggest potential solutions
- Point to relevant documentation

Would you be interested in working on this?

**Comment by vubangsi** [comment]
Thank you for reporting this issue! 

This looks like a great contribution opportunity. I noticed this repository is looking for help with $good first issue, help wanted issues.

As an automated agent, I'm here to help the open source community. This issue seems well-suited for community contributions.

If you'd like, I can help:
- Review the issue requirements
- Suggest potential solutions
- Point to relevant documentation

Would you be interested in working on this?

**Comment by vubangsi** [comment]
Thank you for reporting this issue! 

This looks like a great contribution opportunity. I noticed this repository is looking for help with $good first issue, help wanted issues.

As an automated agent, I'm here to help the open source community. This issue seems well-suited for community contributions.

If you'd like, I can help:
- Review the issue requirements
- Suggest potential solutions
- Point to relevant documentation

Would you be interested in working on this?

**Comment by vubangsi** [comment]
Thank you for reporting this issue! 

This looks like a great contribution opportunity. I noticed this repository is looking for help with $good first issue, help wanted issues.

As an automated agent, I'm here to help the open source community. This issue seems well-suited for community contributions.

If you'd like, I can help:
- Review the issue requirements
- Suggest potential solutions
- Point to relevant documentation

Would you be interested in working on this?

**Comment by MaximSmolskiy** [comment]
@vubangsi stop spamming

**Comment by Shashank0701-byte** [comment]
Hi! I'm interested in working on this issue. Could you please let me know if it's still available or if there are any guidelines I should follow before starting? I'm looking forward to contributing and would appreciate any advice or pointers on how to get started. Thanks!
